GCC build failure for JIT code

I hit some GCC build failure in JIT code in my PR [here](https://github.com/dotnet/runtime/pull/57101), and I don't think it has to do with my changes.

The official build log simply repeated these few failures many times:
```
In file included from /__w/1/s/src/coreclr/jit/compiler.h:30,
                 from /__w/1/s/src/coreclr/jit/jit.h:758,
                 from /__w/1/s/src/coreclr/jit/jitpch.h:18,
                 from /__w/1/s/src/coreclr/jit/alloc.cpp:4:
/__w/1/s/src/coreclr/jit/lir.h:301:30: warning: extra qualification 'LIR::' on member 'AsRange' [-fpermissive]
   301 |     static const LIR::Range& LIR::AsRange(const BasicBlock* block);
       |                              ^~~
 In file included from /__w/1/s/src/coreclr/jit/codegeninterface.h:26,
                  from /__w/1/s/src/coreclr/jit/compiler.h:51,
                  from /__w/1/s/src/coreclr/jit/jit.h:758,
                  from /__w/1/s/src/coreclr/jit/jitpch.h:18,
                  from /__w/1/s/src/coreclr/jit/alloc.cpp:4:
 /__w/1/s/src/coreclr/jit/emit.h:536:25: warning: 'emitter::emitAddrMode::amScale' is too small to hold all values of 'enum emitter::opSize'
   536 |         emitter::opSize amScale : 2;
       |                         ^~~~~~~
 /__w/1/s/src/coreclr/jit/emit.h:855:27: warning: 'emitter::instrDesc::idAddrUnion::<unnamed struct>::_idReg3' is too small to hold all values of 'regNumber' {aka 'enum _regNumber_enum'}
   855 |                 regNumber _idReg3 : REGNUM_BITS;
       |                           ^~~~~~~
 /__w/1/s/src/coreclr/jit/emit.h:856:27: warning: 'emitter::instrDesc::idAddrUnion::<unnamed struct>::_idReg4' is too small to hold all values of 'regNumber' {aka 'enum _regNumber_enum'}
   856 |                 regNumber _idReg4 : REGNUM_BITS;
       |                           ^~~~~~~
 /__w/1/s/src/coreclr/jit/emit.h:669:19: warning: 'emitter::instrDesc::_idReg1' is too small to hold all values of 'regNumber' {aka 'enum _regNumber_enum'}
   669 |         regNumber _idReg1 : REGNUM_BITS; // register num
       |                   ^~~~~~~
 /__w/1/s/src/coreclr/jit/emit.h:671:19: warning: 'emitter::instrDesc::_idReg2' is too small to hold all values of 'regNumber' {aka 'enum _regNumber_enum'}
   671 |         regNumber _idReg2 : REGNUM_BITS;
```
The warning on line 301 is obvious. All we needed to do is to get rid of the extra qualification.

The warning on line 536 is valid, there are 7 distinct values for `opSize`, but we are allocating only 2 bits for `amScale`.
https://github.com/dotnet/runtime/blob/c5f5b99ae1f084b802a91bf632b3e784cea7bb1e/src/coreclr/jit/emit.h#L457-L471

It is unclear what's wrong with the `REGNUM_BITS`. This is a failure on x64, and on x64 we should have 36 distinct enum values, and we gave 6 bits, that should be enough, but somehow GCC complains.

```
0:000> dt clrjit!emitter::instrDesc
   +0x000 _idIns           : Pos 0, 10 Bits
   +0x000 _idInsFmt        : Pos 10, 7 Bits
   +0x000 _idCodeSize      : Pos 17, 4 Bits
   +0x000 _idOpSize        : Pos 21, 3 Bits
   +0x000 _idGCref         : Pos 24, 2 Bits
   +0x000 _idReg1          : Pos 26, 6 Bits
   +0x004 _idReg2          : Pos 0, 6 Bits
   +0x004 _idSmallDsc      : Pos 6, 1 Bit
   +0x004 _idLargeCns      : Pos 7, 1 Bit
   +0x004 _idLargeDsp      : Pos 8, 1 Bit
   +0x004 _idLargeCall     : Pos 9, 1 Bit
   +0x004 _idBound         : Pos 10, 1 Bit
   +0x004 _idCallRegPtr    : Pos 11, 1 Bit
   +0x004 _idCallAddr      : Pos 12, 1 Bit
   +0x004 _idNoGC          : Pos 13, 1 Bit
   +0x004 _idCnsReloc      : Pos 14, 1 Bit
   +0x004 _idDspReloc      : Pos 15, 1 Bit
   +0x004 _idSmallCns      : Pos 16, 16 Bits
   +0x008 _idDebugOnlyInfo : Ptr64 emitter::instrDescDebugInfo
   +0x010 _idAddrUnion     : emitter::instrDesc::idAddrUnion
0:000> dt clrjit!_regNumber_enum
   REG_RAX = 0n0
   REG_RCX = 0n1
   REG_RDX = 0n2
   REG_RBX = 0n3
   REG_RSP = 0n4
   REG_RBP = 0n5
   REG_RSI = 0n6
   REG_RDI = 0n7
   REG_R8 = 0n8
   REG_R9 = 0n9
   REG_R10 = 0n10
   REG_R11 = 0n11
   REG_R12 = 0n12
   REG_R13 = 0n13
   REG_R14 = 0n14
   REG_R15 = 0n15
   REG_EAX = 0n0
   REG_ECX = 0n1
   REG_EDX = 0n2
   REG_EBX = 0n3
   REG_ESP = 0n4
   REG_EBP = 0n5
   REG_ESI = 0n6
   REG_EDI = 0n7
   REG_XMM0 = 0n16
   REG_XMM1 = 0n17
   REG_XMM2 = 0n18
   REG_XMM3 = 0n19
   REG_XMM4 = 0n20
   REG_XMM5 = 0n21
   REG_XMM6 = 0n22
   REG_XMM7 = 0n23
   REG_XMM8 = 0n24
   REG_XMM9 = 0n25
   REG_XMM10 = 0n26
   REG_XMM11 = 0n27
   REG_XMM12 = 0n28
   REG_XMM13 = 0n29
   REG_XMM14 = 0n30
   REG_XMM15 = 0n31
   REG_STK = 0n32
   REG_COUNT = 0n33
   REG_NA = 0n33
   ACTUAL_REG_COUNT = 0n32
```

This [post](https://stackoverflow.com/questions/36005063/gcc-suppress-warning-too-small-to-hold-all-values-of) on StackOverflow seems to imply that the warning is independent on the enum values. If that's the case then it at least makes sense, but I wonder what should we do about that, we can't just leave the warning there.

error is:

```
2021-08-20T01:29:54.2868927Z /__w/1/s/src/coreclr/gc/gc.cpp:43216:24: error: comparison of integer expressions of different signedness: 'unsigned int' and 'int' [-Werror=sign-compare]
2021-08-20T01:29:54.2870382Z 43216 |         if (generation == -1)
2021-08-20T01:29:54.2871150Z       |             ~~~~~~~~~~~^~~~~
```

--- a/src/coreclr/jit/CMakeLists.txt
+++ b/src/coreclr/jit/CMakeLists.txt
@@ -4,8 +4,7 @@ include_directories("./jitstd")
 include_directories("../inc")
 
 if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
-  add_compile_options($<$<COMPILE_LANGUAGE:CXX>:-fpermissive>)
-  add_compile_options(-Wno-error)
+  add_compile_options(-Wno-error) # tracked by https://github.com/dotnet/runtime/issues/33541
 endif()
 
 function(create_standalone_jit)

--- a/src/coreclr/jit/bitsetasshortlong.h
+++ b/src/coreclr/jit/bitsetasshortlong.h
@@ -345,7 +345,7 @@ public:
     {
         if (IsShort(env))
         {
-            (size_t&)out = (size_t)out & ((size_t)gen | (size_t)in);
+            (size_t&)(int&) out = (size_t)out & ((size_t)gen | (size_t)in);
         }
         else
         {
@@ -361,7 +361,7 @@ public:
     {
         if (IsShort(env))
         {
-            (size_t&)in = (size_t)use | ((size_t)out & ~(size_t)def);
+            (size_t&)(int&) in = (size_t)use | ((size_t)out & ~(size_t)def);
         }
         else
         {

--- a/src/coreclr/jit/codegenxarch.cpp
+++ b/src/coreclr/jit/codegenxarch.cpp
@@ -79,7 +79,7 @@ void CodeGen::genSetGSSecurityCookie(regNumber initReg, bool* pInitRegZeroed)
     {
         noway_assert(compiler->gsGlobalSecurityCookieVal != 0);
 #ifdef TARGET_AMD64
-        if ((int)compiler->gsGlobalSecurityCookieVal != compiler->gsGlobalSecurityCookieVal)
+        if ((size_t)(int)compiler->gsGlobalSecurityCookieVal != compiler->gsGlobalSecurityCookieVal)
         {
             // initReg = #GlobalSecurityCookieVal64; [frame.GSSecurityCookie] = initReg
             genSetRegToIcon(initReg, compiler->gsGlobalSecurityCookieVal, TYP_I_IMPL);
@@ -2511,7 +2511,7 @@ void CodeGen::genLclHeap(GenTree* tree)
             amount /= STACK_ALIGN;
         }
 
-        genSetRegToIcon(regCnt, amount, ((int)amount == amount) ? TYP_INT : TYP_LONG);
+        genSetRegToIcon(regCnt, amount, ((size_t)(int)amount == amount) ? TYP_INT : TYP_LONG);
     }
 
     if (compiler->info.compInitMem)

--- a/src/coreclr/jit/emitarm.cpp
+++ b/src/coreclr/jit/emitarm.cpp
@@ -6790,7 +6790,7 @@ void emitter::emitDispImm(int imm, bool addComma, bool alwaysHex /* =false */)
         printf("%d", imm);
     else if ((imm > 0) ||
              (imm == -imm) || // -0x80000000 == 0x80000000. So we don't want to add an extra "-" at the beginning.
-             (emitComp->opts.disDiffable && (imm == 0xD1FFAB1E))) // Don't display this as negative
+             (emitComp->opts.disDiffable && (imm == (int)0xD1FFAB1E))) // Don't display this as negative
         printf("0x%02x", imm);
     else // val <= -1000
         printf("-0x%02x", -imm);

--- a/src/coreclr/jit/emitxarch.cpp
+++ b/src/coreclr/jit/emitxarch.cpp
@@ -7624,7 +7624,7 @@ void emitter::emitIns_Call(EmitCallType          callType,
                         // An absolute indir address that doesn't need reloc should fit within 32-bits
                         // to be encoded as offset relative to zero.  This addr mode requires an extra
                         // SIB byte
-                        noway_assert(static_cast<int>(reinterpret_cast<intptr_t>(addr)) == (size_t)addr);
+                        noway_assert((size_t) static_cast<int>(reinterpret_cast<intptr_t>(addr)) == (size_t)addr);
                         sz++;
                     }
 #endif // TARGET_AMD64
@@ -7660,7 +7660,7 @@ void emitter::emitIns_Call(EmitCallType          callType,
             // An absolute indir address that doesn't need reloc should fit within 32-bits
             // to be encoded as offset relative to zero.  This addr mode requires an extra
             // SIB byte
-            noway_assert(static_cast<int>(reinterpret_cast<intptr_t>(addr)) == (size_t)addr);
+            noway_assert((size_t) static_cast<int>(reinterpret_cast<intptr_t>(addr)) == (size_t)addr);
             sz++;
         }
 #endif // TARGET_AMD64

--- a/src/coreclr/jit/gcencode.cpp
+++ b/src/coreclr/jit/gcencode.cpp
@@ -1672,7 +1672,7 @@ size_t GCInfo::gcInfoBlockHdrSave(
 #if REGEN_SHORTCUTS
             regenLog(headerEncoding, header, &state);
 #endif
-            _ASSERTE(codeSet == 1 || codeSet == 2 && "Encoding must correspond to InfoHdrAdjust or InfoHdrAdjust2");
+            _ASSERTE((codeSet == 1 || codeSet == 2) && "Encoding must correspond to InfoHdrAdjust or InfoHdrAdjust2");
             if (codeSet == 2)
             {
                 *dest++ = NEXT_OPCODE | MORE_BYTES_TO_FOLLOW;

--- a/src/coreclr/jit/inlinepolicy.h
+++ b/src/coreclr/jit/inlinepolicy.h
@@ -185,7 +185,7 @@ protected:
 class ExtendedDefaultPolicy : public DefaultPolicy
 {
 public:
-    ExtendedDefaultPolicy::ExtendedDefaultPolicy(Compiler* compiler, bool isPrejitRoot)
+    ExtendedDefaultPolicy(Compiler* compiler, bool isPrejitRoot)
         : DefaultPolicy(compiler, isPrejitRoot)
         , m_ProfileFrequency(0.0)
         , m_BinaryExprWithCns(0)

--- a/src/coreclr/jit/lir.h
+++ b/src/coreclr/jit/lir.h
@@ -298,7 +298,7 @@ public:
 
 public:
     static Range& AsRange(BasicBlock* block);
-    static const LIR::Range& LIR::AsRange(const BasicBlock* block);
+    static const Range& AsRange(const BasicBlock* block);
 
     static Range EmptyRange();
     static Range SeqTree(Compiler* compiler, GenTree* tree);

--- a/src/coreclr/jit/morph.cpp
+++ b/src/coreclr/jit/morph.cpp
@@ -3394,7 +3394,7 @@ void Compiler::fgInitArgInfo(GenTreeCall* call)
                     if (callIsVararg)
                     {
                         // Override the decision and force a split.
-                        isRegArg = isRegArg = (intArgRegNum + (size - 1)) <= maxRegArgs;
+                        isRegArg = (intArgRegNum + (size - 1)) <= maxRegArgs;
                     }
                     else
 #endif // defined(TARGET_WINDOWS)

